x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/UnigramTokenizer.java

```diff
@@ -0,0 +1,489 @@
+            List<BytesRef> prefixes = new ArrayList<>();
+            int numMatchedChildren = 0;
+            BytesTrie node = this;
+            for (int i = input.offset; i < input.length + input.offset; i++) {
+                if (node == null) {
+                    break;
+                }
+                if (node.isLeaf() && numMatchedChildren > 0) {
+                    prefixes.add(new BytesRef(input.bytes, input.offset, numMatchedChildren));
+                }
+                node = node.children.get(input.bytes[i]);
+                numMatchedChildren++;
+            }
+            if (node != null && node.isLeaf() && numMatchedChildren > 0) {
+                prefixes.add(new BytesRef(input.bytes, input.offset, numMatchedChildren));
+            }
+            return prefixes;
+        }
+
+        void insert(BytesRef bytes) {
+            if (bytes.length == 0) {
+                return;
+            }
+            BytesTrie currentNode = this;
+            int currentTokenIndex = 0;
+
+            // find last child
+            while (currentTokenIndex < bytes.length) {
+                currentNode = currentNode.children.computeIfAbsent(fromBytesRef(bytes, currentTokenIndex), k -> new BytesTrie());
+                currentTokenIndex++;
+            }
+            currentNode.setLeaf(true);
+        }
+
+        public static BytesTrie build(Collection<BytesRef> tokens) {
+            BytesTrie root = new BytesTrie();
+            for (BytesRef token : tokens) {
+                root.insert(token);
+            }
+            return root;
+        }
+    }
+
+    /**
+     * This keeps track of the best-path in the vocab for given prefixes
+     */
+    private static class BestPathNode {
+        // Token Id, -1 if its unknown
+        private int id = -1;
+        // Token score
+        double score = 0.0;
+        // starts at byte position for walking back the best scoring node
+        private int startsAtBytePos = -1;
+        // Its char position for correctly identifying offsets related to the original input
+        private int startsAtCharPos = -1;
+    }
+
+    @FunctionalInterface
+    public interface IntToIntFunction {
+        int apply(int value);
+    }
+
+    /**
+     * This is a simple whitespace tokenizer that generates whitespace delimited tokens from the input stream
+     *
+     * This is effectively the lucene WhitespaceTokenizer, slightly adjusted for our needs here.
+     */
+    class SimpleWhitespaceTokenizer {
+        private int offset = 0, bufferIndex = 0, dataLen = 0, finalOffset = 0;
+        private static final int IO_BUFFER_SIZE = 4096;
+        private final CharacterUtils.CharacterBuffer ioBuffer = CharacterUtils.newCharacterBuffer(IO_BUFFER_SIZE);
+
+        void reset() {
+            bufferIndex = 0;
+            offset = 0;
+            dataLen = 0;
+            finalOffset = 0;
+            ioBuffer.reset();
+        }
+
+        Optional<DelimitedToken> next() throws IOException {
```

Comment:
It's not necessary to change it for this PR, but I think there could be a cost to using `Optional` in loops where we expect to create and destroy many `Optional` objects very quickly.  For example see https://stackoverflow.com/a/42179644.  As a followup it's probably worth checking if we can improve our tokenization performance by switching from `Optional` to `null` in the inner loops.